modules/samples/watertank/component/api/src/main/java/org/homeautomation/watertank/api/DeviceTypeServiceImpl.java

```diff
@@ -0,0 +1,434 @@
+            return Response.ok().build();
+        } catch (DeviceAccessAuthorizationException e) {
+            log.error(e.getErrorMessage(), e);
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR).build();
+        }
+    }
+
+    /**
+     * Retrieve Sensor data for the given time period
+     *
+     * @param deviceId   unique identifier for given device type instance
+     * @param sensorName name of the sensor
+     * @param from       starting time
+     * @param to         ending time
+     * @return response with List<SensorRecord> object which includes sensor data which is requested
+     */
+    @Path("device/stats/{deviceId}/sensors/{sensorName}")
+    @GET
+    @Consumes("application/json")
+    @Produces("application/json")
+    public Response getSensorStats(@PathParam("deviceId") String deviceId, @PathParam("sensorName") String sensorName,
+                                   @QueryParam("from") long from, @QueryParam("to") long to) {
+        String fromDate = String.valueOf(from);
+        String toDate = String.valueOf(to);
+        String query = "deviceId:" + deviceId + " AND deviceType:" +
+                       DeviceTypeConstants.DEVICE_TYPE + " AND time : [" + fromDate + " TO " + toDate + "]";
+        String sensorTableName;
+        switch (sensorName) {
+            case DeviceTypeConstants.STREAM_RELAY:
+                sensorTableName = DeviceTypeConstants.RELAY_EVENT_TABLE;
+                break;
+            case DeviceTypeConstants.STREAM_WATERLEVEL:
+                sensorTableName = DeviceTypeConstants.WATERLEVEL_EVENT_TABLE;
+                break;
+            default:
+                return Response.status(Response.Status.BAD_REQUEST).entity("Invalid event stream").build();
+        }
+
+        try {
+            if (!APIUtil.getDeviceAccessAuthorizationService().isUserAuthorized(new DeviceIdentifier(deviceId,
+                                                                                                     DeviceTypeConstants.DEVICE_TYPE))) {
+                return Response.status(Response.Status.UNAUTHORIZED.getStatusCode()).build();
+            }
+            List<SortByField> sortByFields = new ArrayList<>();
+            SortByField sortByField = new SortByField("time", SORT.ASC, false);
+            sortByFields.add(sortByField);
+            List<SensorRecord> sensorRecords = APIUtil.getAllEventsForDevice(sensorTableName, query, sortByFields);
+            return Response.status(Response.Status.OK.getStatusCode()).entity(sensorRecords).build();
+        } catch (AnalyticsException e) {
+            String errorMsg = "Error on retrieving stats on table " + sensorTableName + " with query " + query;
+            log.error(errorMsg);
+            return Response.status(Response.Status.INTERNAL_SERVER_ERROR.getStatusCode()).entity(errorMsg).build();
+        } catch (DeviceAccessAuthorizationException e) {
+            log.error(e.getErrorMessage(), e);
```

Comment:
DeviceAccessAuthorizationException is a custom exception which was in CDMF. It is already providing underlying error message which is relevant to the error. So why we need to re write the exception message again other than showing the actual exception. As logger printing exception with the class name we can know it is occurring in here.